### PR TITLE
PR Modulo 1

### PR DESCRIPTION
**Pull request by vzadri**
La función para validar el esquema del json al final está comentada su llamada, pero la incluyo ya que para las pruebas con poco datos sacados de la API en principio funcionaba, pero al final al llamar a la API con todos los años daba un error en el esquema que no he podido solucionar por el rate limit de la API.

### CODE REVIEW

File: src/module_1/module_1_meteo_api.py [hunk 1]
```diff
@@ -1,7 +1,311 @@
...
+    "FGOALS_f3_H",
+    "HiRAM_SIT_HR",
+    "MRI_AGCM3_2_S",
+    "EC_Earth3P_HR",
+    "MPI_ESM1_2_XR",
+    "NICAM16_8S",
+)
+
+
+def validate_json_schema(response):
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:24]: I suggedt you to use type hinting

File: src/module_1/module_1_meteo_api.py [hunk 2]
```diff
@@ -1,7 +1,311 @@
...
+            "longitude",
+            "timezone",
+            "timezone_abbreviation",
+            "utc_offset_seconds",
+        ],
+    }
+    try:
+        jsonschema.validate(response, schema)
+    except jsonschema.exceptions.ValidationError as error:
+        print("Data does not match schema: ", error)
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:217]: Use logging

File: src/module_1/module_1_meteo_api.py [hunk 3]
```diff
@@ -1,7 +1,311 @@
...
+    df_new["temperature_std"] = df_grouped[temp_cols].std().mean(axis=1)
+    df_new["precipitation_mean"] = df_grouped[prec_cols].mean().mean(axis=1)
+    df_new["precipitation_std"] = df_grouped[prec_cols].std().mean(axis=1)
+    df_new["soil_moisture_mean"] = df_grouped[soil_cols].mean().mean(axis=1)
+    df_new["soil_moisture_std"] = df_grouped[soil_cols].std().mean(axis=1)
+
+    return df_new
+
+
+def plot_meteo_data(dict_city_df):
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:285]: Docstring here would help a lot. What is the expected entry in terms of columns and data?

File: src/module_1/module_1_meteo_api.py [hunk 4]
```diff
@@ -1,7 +1,311 @@
...
+        for i, variable in enumerate(["temperature", "precipitation", "soil_moisture"]):
+            mean = df[f"{variable}_mean"]
+            std = df[f"{variable}_std"]
+            axs[i].plot(df.index, mean, label=f"{city} {variable}")
+            axs[i].fill_between(df.index, mean - std, mean + std, alpha=0.2)
+            axs[i].set_ylabel(variable)
+            axs[i].legend()
+
+    plt.tight_layout()
+    plt.show()
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:298]: TIP: If you return the fig and ax will make this method more versatile. Whoever calling it could show the plot or extend it.

File: src/module_1/module_1_meteo_api.py [hunk 5]
```diff
@@ -1,7 +1,311 @@
...
+
+
+def plot_meteo_data(dict_city_df):
+    fig, axs = plt.subplots(3, 1, figsize=(10, 15))
+
+    for city, df in dict_city_df.items():
+        for i, variable in enumerate(["temperature", "precipitation", "soil_moisture"]):
+            mean = df[f"{variable}_mean"]
+            std = df[f"{variable}_std"]
+            axs[i].plot(df.index, mean, label=f"{city} {variable}")
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:292]: Makes more sense for the label to only have the city. The variable can go in the title

File: src/module_1/module_1_meteo_api.py [hunk 6]
```diff
@@ -1,7 +1,311 @@
...
+            "models": MODELS,
+            "daily": VARIABLES,
+        }
+        data = call_api(API_URL, params)
+        data = pd.DataFrame(data)
+        return data
+
+
+def process_dataframe_columns(df):
+    df["time"] = pd.to_datetime(df["time"]).dt.year
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:254]: Not the best variable naming. Year would be a better choice

File: src/module_1/module_1_meteo_api.py [hunk 7]
```diff
@@ -1,7 +1,311 @@
...
+        print("error=", error)
+    else:
+        response = response.json()
+        # if validate_json_schema(response):
+        print("json schema correcto")
+        return response["daily"]
+
+
+def get_data_meteo_api(city):
+    if city in COORDINATES:
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:238]: You are not dealing with the false condition

File: src/module_1/module_1_meteo_api.py [hunk 8]
```diff
@@ -1,7 +1,311 @@
...
+        params = {
+            "latitude": COORDINATES[city]["latitude"],
+            "longitude": COORDINATES[city]["longitude"],
+            "start_date": "1950-01-01",
+            # "end_date": "2050-12-31",
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:243]: Remove line

File: src/module_1/module_1_meteo_api.py [hunk 9]
```diff
@@ -1,7 +1,311 @@
...
+    url,
+    params,
+):
+    response = requests.get(url, params=params)
+    if not response.status_code == 200:
+        error = response.json()
+        print("error=", error)
+    else:
+        response = response.json()
+        # if validate_json_schema(response):
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:232]: Why commenting this?

File: src/module_1/module_1_meteo_api.py [hunk 10]
```diff
@@ -1,7 +1,311 @@
...
+    }
+    try:
+        jsonschema.validate(response, schema)
+    except jsonschema.exceptions.ValidationError as error:
+        print("Data does not match schema: ", error)
+        return False
+    return True
+
+
+def call_api(
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:222]: This, while functioning, it is very simple. I recommend you to check the solution class. But some outlines:

- what if you exceed the quota? Think on strategies of retrial
- there may be different errors you can handle from the request

File: src/module_1/module_1_meteo_api.py [hunk 11]
```diff
@@ -1,7 +1,311 @@
...
+    "HiRAM_SIT_HR",
+    "MRI_AGCM3_2_S",
+    "EC_Earth3P_HR",
+    "MPI_ESM1_2_XR",
+    "NICAM16_8S",
+)
+
+
+def validate_json_schema(response):
+    schema = {
```
Comment (pabaldonedo) [inline src/module_1/module_1_meteo_api.py:25]: Just in case you do not know it, you may be interested in checking pydantic for these kind of checks

File: tests/module_1/test_meteo_api.py [hunk 12]
```diff
@@ -1,7 +1,53 @@
-""" This is a dummy example to show how to import code from src/ for testing"""
+import pandas as pd
```
Comment (pabaldonedo) [inline tests/module_1/test_meteo_api.py:1]: I suggest you to have a look at pytest packag, the use of fixtures for data, and the Mock object of unittest package

File: tests/module_1/test_meteo_api.py [hunk 13]
```diff
@@ -1,7 +1,53 @@
...
+
+
+def test_get_data_meteo_api():
+    with requests_mock.Mocker() as mock:
+        mock.get(API_URL, json=mock_data)
+        data = get_data_meteo_api("Madrid")
+
+        assert isinstance(data, pd.DataFrame)
+
+        assert not data.empty
```
Comment (pabaldonedo) [inline tests/module_1/test_meteo_api.py:31]: You can makr use of pandas.testing.assert_frame_equal

File: tests/module_1/test_meteo_api.py [hunk 14]
```diff
@@ -1,7 +1,53 @@
...
+
+
+def test_process_dataframe():
```
Comment (pabaldonedo) [inline tests/module_1/test_meteo_api.py:34]: This is not a unittest, it is more like an integration test. I suggest you to only test process_dataframe method and you provide directly the input without calling get_dara_meteo_api

**Review by pabaldonedo — APPROVED** [review]
Nice first PR!
Ive left some comments on things you can improve:

- Use of type hinting and docstrings
- Better handling the api requests corner cases
- Unittest only tests one thing
- Also some suggestions on packages you can have a look at.
